README.md
# Pathshare SDK for Android


![GitHub release](https://img.shields.io/github/release/freshbits/pathshare-sdk-android.svg?style=flat)
![Platform](https://img.shields.io/badge/platform-android-green.svg?style=flat)
![Language](https://img.shields.io/badge/language-java-brightgreen.svg?style=flat)
![Language](https://img.shields.io/badge/language-kotlin-brightgreen.svg?style=flat)

**Pathshare** is a realtime location sharing platform. For more information please visit the [Pathshare Developer Page](https://pathsha.re/professional/developers).

<img src="/assets/android-example-app.png" height="600">

- [Requirements](#requirements)
- [Installation](#installation)
- [Basic Usage](#basic-usage)
  - [Initialize Pathshare](#init-pathshare)
  - [Save Username](#save-username)
  - [Create Session](#create-session)
  - [Join Session](#join-session)
  - [Invite Customer](#invite-customer)
  - [Leave Session](#leave-session)
  - [Find Session](#find-session)

## Requirements

`PathshareSDK` for Android supports Android 5.x, 6.x, 7.x, 8.x, 9.x, 10.x, 11.x, 12.x and 13.x.

## Installation

The installation of the `PathshareSDK` is simple. Download the latest `pathshare-sdk-android-[version].zip` from [Releases](https://github.com/freshbits/pathshare-sdk-android/releases), unzip and copy the `PathshareSDK` folder into the root of your project.

Next, reference the `PathareSDK` folder in your application `build.gradle` file:

```gradle
repositories {
    maven { url "file://$projectDir/../PathshareSDK" }
}

dependencies {
    ...
    compile 'ch.freshbits.pathshare.sdk:pathshare-sdk:2.3.3'
}
```

## Basic Usage

### Initialization

In order to initialize the `PathshareSDK`, create a file named `pathshare.xml` inside the `res/value` folder and add your account token:

```xml
<?xml version="1.0" encoding="utf-8"?>
<resources>
    <string name="pathshare_account_token">your Pathshare account token</string>
</resources>
```

Next, add the following to the `onCreate` method of your `Application` class:

###### Java
```java
public class ExampleApplication extends Application {
    @Override
    public void onCreate() {
        super.onCreate();

        Pathshare.initialize(this, getString(R.string.pathshare_account_token), TrackingMode.SMART);
    }
}
```
###### Kotlin
```kotlin
class ExampleApplication: Application() {

    override fun onCreate() {
        super.onCreate()

        Pathshare.initialize(this, getString(R.string.pathshare_account_token), TrackingMode.SMART)
    }
}
```

Optionally, you can specify a tracking mode to configure the behavior of the location tracker. The following tracking modes are available:

Tracking Mode      | Description
-------------------|------------------------------------------------------------
`SMART`            | Adapts intelligently to the environment and usage of the device. Includes awareness of battery level, travel speed and motion activity.
`ECO`              | Static mode providing constant tracking data with very low accuracy (several kilometers) and great distance between single locations and ensuring maximum battery life.
`APPROXIMATE`      | Static mode providing constant tracking data with low accuracy (several hundred meters) and middle distance between single locations. Useful when a low battery drain is a important criteria.
`ACCURATE`         | Static mode providing constant tracking with the highest accuracy possible (few meters) and small distances between locations. Useful for scenarios where a high accuracy is an essential requirement.

### Save User

Before creating a session, you need to save a user:

###### Java
```java
Pathshare.client().saveUser("Candice", "me@email.com", "+12345678901", UserType.DRIVER, getResources().getDrawable(R.drawable.face, null), new ResponseListener() {
    @Override
    public void onSuccess() {
        // ...
    }

    @Override
    public void onError(@NonNull Throwable throwable) {
        // ...
    }
});
```

###### Kotlin
```kotlin
Pathshare.client().saveUser("SDK User Android", "me@email.com", "+12345678901", UserType.DRIVER, resources.getDrawable(R.drawable.face, null), object: ResponseListener {
    override fun onSuccess() {
        // ...
    }

    override fun onError(throwable: Throwable) {
        // ...
    }
})
```
The email address can be `null`.

Use the same method `Pathshare.client().saveUser()` to create or update the user.

There are different types of users for specific industries:

User Types                  | Description
----------------------------|------------------------------------------------------------
`TECHNICIAN`, `MOTORIST`    | For roadside assitance industry or similar
`DRIVER`, `RECIPIENT`       | For delivery services or similar
`INVESTIGATOR`, `CLIENT`    | For legal services industry or similar

### Create Session

To create a session, use the session builder:

###### Java
```java
session = new Session.Builder()
    .setExpirationDate(expirationDate)
    .setName("Shopping")
    .build();
```

###### Kotlin
```kotlin
session = Session.Builder()
    .setExpirationDate(expirationDate)
    .setName("Shopping")
    .build()
```

A session needs an expiration date and a name. You can create multiple sessions at the same time, the SDK will manage them for you.


Make sure to save the session after building:

###### Java
```java
session.save(new ResponseListener() { ... });

session.getIdentifier() // => 3fd919fe824d8e7b78e2c11c1570a6f168d2c...
session.isExpired() // => false
session.getURL() // => https://pathsha.re/6d39d5
```

###### Kotlin
```kotlin
session.save(object: ResponseListener { ... })

session.identifier // => 3fd919fe824d8e7b78e2c11c1570a6f168d2c...
session.isExpired // => false
session.url // => https://pathsha.re/6d39d5
```

#### Expiration

In order to react to the expiration of the session, add an `ExpirationListener`:

###### Java
```java
session = new Session.Builder()
    // ...
    .setExpirationListener(new SessionExpirationListener() {
        @Override
        public void onExpiration() {
            // ...
        }
    })
    .build();
```

###### Kotlin
```kotlin
session = Session.Builder()
    // ...
    .setSessionExpirationListener { // ... }
    .build()
```

#### Destination

Optionally, you can add a destination to the session. Sessions with destination will show the estimated time of arrival (ETA) for each user. The destination identifier is used to group sessions by destination.

###### Java
```java
Destination destination = new Destination.Builder()
    .setIdentifier("W2342")
    .setLatitude(47.378178)
    .setLongitude(8.539256)
    .build();

session = new Session.Builder()
    // ...
    .setDestination(destination)
    .build();
```

###### Kotlin
```kotlin
val destination = Destination.Builder()
    .setIdentifier("W2342")
    .setLatitude(47.378178)
    .setLongitude(8.539256)
    .build()

session = Session.Builder()
    // ...
    .setDestination(destination)
    .build()
```

### Join Session

To join the session you created, call the `joinUser()` method on the session object:

###### Java
```java
session.join(new ResponseListener() { ... });

session.isUserJoined() // => true
```

###### Kotlin
```kotlin
session.join(object: ResponseListener { ... })

session.isUserJoined // => true
```

This call will add your Pathshare user to the session and you will be able to see his location on a map in realtime in the Pathshare Professional web interface.

### Invite customer

To invite a customer to the session, call the `inviteUser()` method on the session object:

###### Java
```java
session.inviteUser("Customer", UserType.CLIENT, "customer@email.com", "+12345678901", new InvitationResponseListener() {
    @Override
    public void onSuccess(URL url) {
        // ...
        Log.d("URL", url.toString()); // => https://m.pathsha.re/12s83a
    }

    public void onError(@NonNull Throwable throwable) {
        // ...
    }
});
```

###### Kotlin
```kotlin
session.inviteUser("Customer", UserType.CLIENT, "customer@me.com", "+12345678901", object: InvitationResponseListener {
    override fun onSuccess(url: URL?) {
        // ...
        Log.d("URL", url.toString()) // => https://m.pathsha.re/12s83a
    }

    override fun onError(throwable: Throwable) {
        // ...
    }
})
```

This call will create a customer user and return an invitation URL that can be sent to the customer using your preffered channel. The customer will then see the driver's location in realtime as well as the ETA in a white-labeled view with your corporate identity.

The customer will be able to enjoy the full realtime experience in the web browser of their smartphone:

<img src="/assets/web-view-customer.png" height="600">

### Leave Session

In order to stop sending user location and remove the user from the session, call the `leaveUser()` method:

###### Java
```java
session.leave(new ResponseListener() { ... });
```

###### Kotlin
```kotlin
session.leave(object: ResponseListener { ... })
```

### Find Session

To find an existing session, use the `findSession()` method with the corresponding session identifier:

###### Java
```java
Pathshare.client().findSession(identifier, new SessionResponseListener() {
    @Override
    public void onSuccess(Session session) {
        session.setSessionExpirationListener(
            new SessionExpirationListener() { ... }
        );
    }

    @Override
    public void onError(@NonNull Throwable throwable) { ... }
}
```

###### Kotlin
```kotlin
Pathshare.client().findSession(identifier, object: SessionResponseListener {
    override fun onSuccess(session: Session?) {
        session.sessionExpirationListener = SessionExpirationListener { ... }
    }

    override fun onError(throwable: Throwable) { ... }
})
```


CHANGELOG.md
# Change Log

## 2.3.3 - 2023-02-26
### Changed
- Updated to support compileSdkVersion to 31
- Minor improvements

## 2.3.2 - 2020-04-03
### Changed
- Added better error handling

## 2.3.1 - 2020-01-31
### Changed
- Added manufacturer specific support
- Minor improvements

## 2.3.0 - 2018-12-01
### Changed
- Added image support to user
- Added support to define if customer can update session destination
- Minor improvements

## 2.2.0 - 2018-09-10
### Changed
- Added email address to user
- Added support for null values for email address
- Minor improvements

## 2.1.0 - 2018-08-30
### Changed
- Moved tracking mode configuration to the Pathshare client to use the same config across all sessions
- Add support for multi-session
- Minor improvements

## 2.0.0 - 2018-08-20
### Changed
- Updated to use our enhanced API
- Added possibility to invite customer and get a inique inivation URL back
- Minor improvements


## 1.0.1 - 2017-08-08
### Changed
- Added support for Android 7.x.
- Minor improvements.

## 1.0.0 - 2015-12-23
### Added
- Pathshare SDK for Android.


android-example-app/app-kotlin/src/main/java/ch/freshbits/pathshare/example/MainActivity.kt
package ch.freshbits.pathshare.example

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import ch.freshbits.pathshare.sdk.Pathshare
import ch.freshbits.pathshare.sdk.helper.*
import ch.freshbits.pathshare.sdk.model.Destination
import ch.freshbits.pathshare.sdk.model.Session
import ch.freshbits.pathshare.sdk.model.UserType
import java.net.URL
import java.util.*

class MainActivity : AppCompatActivity() {

    companion object {
        private const val TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS = 1
        private const val SESSION_PREFERENCES = "session"
        private const val SESSION_ID_KEY = "session_id"
    }

    private lateinit var createButton: Button
    private lateinit var joinButton: Button
    private lateinit var inviteButton: Button
    private lateinit var leaveButton: Button

    private lateinit var session: Session

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initializeCreateButton()
        initializeJoinButton()
        initializeInviteButton()
        initializeLeaveButton()

        findSession()
    }

    private fun initializeCreateButton() {
        createButton = findViewById<View>(R.id.create_session) as Button
        createButton.isEnabled = true
        createButton.setOnClickListener {
            Pathshare.client().saveUser(
                    "SDK User Android",
                    "me@email.com",
                    "+12345678901",
                    UserType.TECHNICIAN,
                    resources.getDrawable(R.drawable.face, null),
                    object: ResponseListener {
                override fun onSuccess() {
                    Log.d("User", "Success")
                    createSession()
                }

                override fun onError(throwable: Throwable) {
                    Log.e("User", "Error: ${throwable.localizedMessage}")
                }
            })
        }
    }

    private fun initializeJoinButton() {
        joinButton = findViewById<View>(R.id.join_session) as Button
        joinButton.isEnabled = false
        joinButton.setOnClickListener { joinSession() }
    }

    private fun initializeInviteButton() {
        inviteButton = findViewById<View>(R.id.invite_customer) as Button
        inviteButton.isEnabled = false
        inviteButton.setOnClickListener { inviteCustomer() }
    }

    private fun initializeLeaveButton() {
        leaveButton = findViewById<View>(R.id.leave_session) as Button
        leaveButton.isEnabled = false
        leaveButton.setOnClickListener { leaveSession() }
    }

    private fun findSession() {
        val identifier = preferences().getString(SESSION_ID_KEY, null) ?: return

        Pathshare.client().findSession(identifier, object: SessionResponseListener {
            override fun onSuccess(session: Session?) = if (session == null || session.isExpired) {
                deleteSessionIdentifier()

                createButton.isEnabled = true
                joinButton.isEnabled = false
                inviteButton.isEnabled = false
                leaveButton.isEnabled = false
            } else {
                Log.d("Session", "Name: ${session.name}")
                session.sessionExpirationListener = SessionExpirationListener { handleSessionExpiration() }

                createButton.isEnabled = false
                joinButton.isEnabled = true
                inviteButton.isEnabled = false
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                showToast("Something went wrong")
            }
        })
    }

    private fun createSession() {
        val destination = Destination.Builder()
                .setIdentifier("w9823")
                .setLatitude(37.7875694)
                .setLongitude(-122.4112239)
                .build()

        val expirationDate = Date(System.currentTimeMillis() + 3600000)

        session = Session.Builder()
                .setDestination(destination)
                .setExpirationDate(expirationDate)
                .setName("simple session")
                .setSessionExpirationListener { handleSessionExpiration() }
                .build()

        session.save(object: ResponseListener {
            override fun onSuccess() {
                Log.d("Session", "Success")
                saveSessionIdentifier()

                createButton.isEnabled = false
                joinButton.isEnabled = true
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                Log.e("Session", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun joinSession() {
        if (session.isExpired) return

        if (hasLocationPermission()) {
            performJoinSession()
        } else {
            requestLocationPermission()
        }
    }

    private fun inviteCustomer() {
        if (session.isExpired) return

        session.inviteUser(
                "Customer",
                UserType.MOTORIST,
                "customer@me.com",
                "+12345678901",
                true,
                object: InvitationResponseListener {
            override fun onSuccess(url: URL?) {
                Log.d("Invite", "Success")
                Log.d("URL", url.toString())
                inviteButton.isEnabled = false
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                Log.e("Invite", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun leaveSession() {
        if (session.isExpired) return

        session.leave(object: ResponseListener {
            override fun onSuccess() {
                Log.d("Leave", "Success")
                createButton.isEnabled = true
                joinButton.isEnabled = false
                inviteButton.isEnabled = false
                leaveButton.isEnabled = false

                deleteSessionIdentifier()
            }

            override fun onError(throwable: Throwable) {
                Log.e("Leave", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun requestLocationPermission() {
        PermissionRequester.requestPermission(this, TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS, android.Manifest.permission.ACCESS_FINE_LOCATION, R.string.permission_access_fine_location_rationale)
    }

    private fun performJoinSession() {
        if (session.isUserJoined) return

        session.join(object: ResponseListener {
            override fun onSuccess() {
                Log.d("Join", "Success")
                createButton.isEnabled = false
                joinButton.isEnabled = false
                inviteButton.isEnabled = true
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                Log.e("Join", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun hasLocationPermission(): Boolean {
        return PermissionRequester.hasPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION)
    }

    private fun saveSessionIdentifier() {
        val editor = preferences().edit()
        editor.putString(SESSION_ID_KEY, session.identifier)
        editor.apply()
    }

    private fun handleSessionExpiration() {
        this.runOnUiThread {
            inviteButton.isEnabled = false
            leaveButton.isEnabled = false
            createButton.isEnabled = true
            deleteSessionIdentifier()
            showToast("Session expired")
        }
    }

    private fun showToast(message: String) {
        val toast = Toast.makeText(applicationContext, message, Toast.LENGTH_SHORT)
        toast.show()
    }

    private fun deleteSessionIdentifier() {
        preferences().edit().clear().apply()
    }

    private fun preferences(): SharedPreferences {
        return applicationContext.getSharedPreferences(SESSION_PREFERENCES, MODE_PRIVATE)
    }
}


android-example-app/app/src/main/java/ch/freshbits/pathshare/example/MainActivity.java
package ch.freshbits.pathshare.example;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;

import java.net.URL;
import java.util.Date;

import ch.freshbits.pathshare.sdk.Pathshare;
import ch.freshbits.pathshare.sdk.helper.InvitationResponseListener;
import ch.freshbits.pathshare.sdk.helper.PermissionRequester;
import ch.freshbits.pathshare.sdk.helper.ResponseListener;
import ch.freshbits.pathshare.sdk.helper.SessionResponseListener;
import ch.freshbits.pathshare.sdk.model.Destination;
import ch.freshbits.pathshare.sdk.model.Session;
import ch.freshbits.pathshare.sdk.model.UserType;

public class MainActivity extends Activity {
    private static final int TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS = 1;
    private static final String SESSION_PREFERENCES= "session";
    private static final String SESSION_ID_KEY = "session_id";

    Session mSession;
    Button mCreateButton;
    Button mJoinButton;
    Button mInviteButton;
    Button mLeaveButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initializeCreateButton();
        initializeJoinButton();
        initializeInviteButton();
        initializeLeaveButton();

        findSession();
    }

    private void initializeCreateButton() {
        setCreateButton(findViewById(R.id.create_session));
        getCreateButton().setEnabled(true);
        getCreateButton().setOnClickListener(view -> Pathshare.client().saveUser(
                "SDK User",
                "me@email.com",
                "+12345678901",
                UserType.TECHNICIAN,
                getResources().getDrawable(R.drawable.face, null),
                new ResponseListener() {
            @Override
            public void onSuccess() {
                Log.d("User", "Success");
                createSession();
            }

            @Override
            public void onError(@NonNull Throwable throwable) {
                Log.e("User", "Error: " + throwable.getLocalizedMessage());
            }
        }));
    }

    private void initializeJoinButton() {
        setJoinButton(findViewById(R.id.join_session));
        getJoinButton().setEnabled(false);
        getJoinButton().setOnClickListener(view -> joinSession());
    }

    private void initializeInviteButton() {
        setInviteButton(findViewById(R.id.invite_customer));
        getInviteButton().setEnabled(false);
        getInviteButton().setOnClickListener(view -> inviteCustomer());
    }

    private void initializeLeaveButton() {
        setLeaveButton(findViewById(R.id.leave_session));
        getLeaveButton().setEnabled(false);
        getLeaveButton().setOnClickListener(view -> leaveSession());
    }

    private void createSession() {
        try {
            Destination destination = new Destination.Builder()
                    .setIdentifier("w9823")
                    .setLatitude(37.7875694)
                    .setLongitude(-122.4112239)
                    .build();

            Date expirationDate = new Date(System.currentTimeMillis() + 3600000);

            mSession = new Session.Builder()
                    .setDestination(destination)
                    .setExpirationDate(expirationDate)
                    .setName("simple session")
                    .setSessionExpirationListener(() -> handleSessionExpiration())
                    .build();

            getSession().save(new ResponseListener() {
                @Override
                public void onSuccess() {
                    Log.d("Session", "Success");
                    saveSessionIdentifier();

                    getCreateButton().setEnabled(false);
                    getJoinButton().setEnabled(true);
                    getLeaveButton().setEnabled(true);
                }

                @Override
                public void onError(@NonNull Throwable throwable) {
                    Log.e("Session", "Error: " + throwable.getLocalizedMessage());
                }
            });

        } catch (IllegalArgumentException e) {
            e.printStackTrace();
        }
    }

    private void joinSession() {
        if (getSession().isExpired()) { return; }

        if (hasLocationPermission()) {
            performJoinSession();
        } else {
            requestLocationPermission();
        }
    }

    private void inviteCustomer() {
        if (getSession().isExpired()) { return; }

        getSession().inviteUser(
                "Customer",
                UserType.MOTORIST,
                "customer@me.com",
                "+12345678901",
                true,
                new InvitationResponseListener() {
            @Override
            public void onSuccess(URL url) {
                Log.d("Invite", "Success");
                Log.d("URL", url.toString());
                getInviteButton().setEnabled(false);
                getLeaveButton().setEnabled(true);
            }

            @Override
            public void onError(@NonNull Throwable throwable) {
                Log.e("Invite", "Error: " + throwable.getLocalizedMessage());
            }
        });
    }

    private void requestLocationPermission() {
        PermissionRequester.requestPermission(this, TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS, Manifest.permission.ACCESS_FINE_LOCATION, R.string.permission_access_fine_location_rationale);
    }

    private boolean hasLocationPermission() {
        return PermissionRequester.hasPermission(this, Manifest.permission.ACCESS_FINE_LOCATION);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                performJoinSession();
            } else {
                Toast.makeText(this, R.string.permission_access_fine_location_denied, Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void performJoinSession() {
        try {
            if (getSession().isUserJoined()) { return; }

            getSession().join(new ResponseListener() {
                @Override
                public void onSuccess() {
                    Log.d("Join", "Success");
                    getCreateButton().setEnabled(false);
                    getJoinButton().setEnabled(false);
                    getInviteButton().setEnabled(true);
                    getLeaveButton().setEnabled(true);
                }

                @Override
                public void onError(@NonNull Throwable throwable) {
                    Log.e("Join", "Error: " + throwable.getLocalizedMessage());
                }
            });
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void leaveSession() {
        if (getSession().isExpired()) { return; }

        try {
            getSession().leave(new ResponseListener() {
                @Override
                public void onSuccess() {
                    Log.d("Leave", "Success");
                    getCreateButton().setEnabled(true);
                    getJoinButton().setEnabled(false);
                    getInviteButton().setEnabled(false);
                    getLeaveButton().setEnabled(false);


                    deleteSessionIdentifier();
                }

                @Override
                public void onError(@NonNull Throwable throwable) {
                    Log.e("Leave", "Error: " + throwable.getLocalizedMessage());
                }
            });
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void saveSessionIdentifier() {
        SharedPreferences.Editor editor = getPreferences().edit();
        editor.putString(SESSION_ID_KEY, getSession().getIdentifier());
        editor.apply();
    }

    private void deleteSessionIdentifier() {
        getPreferences().edit().clear().apply();
    }

    private void handleSessionExpiration() {
        this.runOnUiThread(() -> {
            getInviteButton().setEnabled(false);
            getLeaveButton().setEnabled(false);
            getCreateButton().setEnabled(true);
            deleteSessionIdentifier();
            showToast("Session expired.");
        });
    }

    public void setSession(Session session) {
        mSession = session;
    }

    public Session getSession() {
        return mSession;
    }

    public void setCreateButton(Button createButton) {
        this.mCreateButton = createButton;
    }

    public void setJoinButton(Button joinButton) {
        this.mJoinButton = joinButton;
    }

    public void setInviteButton(Button inviteButton) {
        this.mInviteButton = inviteButton;
    }

    public void setLeaveButton(Button leaveButton) {
        this.mLeaveButton = leaveButton;
    }

    public Button getCreateButton() {
        return mCreateButton;
    }

    public Button getJoinButton() {
        return mJoinButton;
    }

    public Button getInviteButton() {
        return mInviteButton;
    }

    public Button getLeaveButton() {
        return mLeaveButton;
    }

    private void findSession() {
        String identifier = getPreferences().getString(SESSION_ID_KEY, null);

        if (identifier == null) { return; }

        Pathshare.client().findSession(identifier, new SessionResponseListener() {
            @Override
            public void onSuccess(Session session) {
                if (session == null || session.isExpired()) {
                    deleteSessionIdentifier();

                    getCreateButton().setEnabled(true);
                    getJoinButton().setEnabled(false);
                    getInviteButton().setEnabled(false);
                    getLeaveButton().setEnabled(false);

                } else {
                    Log.d("Session", "Name: " + session.getName());
                    session.setSessionExpirationListener(() -> handleSessionExpiration());

                    setSession(session);

                    getCreateButton().setEnabled(false);
                    getJoinButton().setEnabled(true);
                    getInviteButton().setEnabled(false);
                    getLeaveButton().setEnabled(true);
                }
            }

            @Override
            public void onError(@NonNull Throwable throwable) {
                showToast("Something went wrong.");
            }
        });
    }

    private void showToast(String message) {
        Toast toast = Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT);
        toast.show();
    }

    private SharedPreferences getPreferences() {
        return getApplicationContext().getSharedPreferences(SESSION_PREFERENCES, Context.MODE_PRIVATE);
    }
}
